packages/dds/tree/src/feature-libraries/default-schema/schemaChecker.ts

```diff
@@ -0,0 +1,105 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	MapTree,
+	StoredSchemaCollection,
+	TreeFieldStoredSchema,
+	TreeNodeStoredSchema,
+	LeafNodeStoredSchema,
+	ObjectNodeStoredSchema,
+	MapNodeStoredSchema,
+} from "../../core/index.js";
+import { FullSchemaPolicy } from "../modular-schema/index.js";
+import { Multiplicity } from "../multiplicity.js";
+import { allowsValue } from "../valueUtilities.js";
+
+export function isNodeInSchema(
+	node: MapTree,
+	schema: TreeNodeStoredSchema,
+	nodeSchemaCollection: StoredSchemaCollection,
+	schemaPolicy: FullSchemaPolicy,
+): boolean {
+	if (
+		schema instanceof LeafNodeStoredSchema &&
+		(node.value === undefined ||
+			node.fields.size !== 0 ||
+			!allowsValue(schema.leafValue, node.value))
+	) {
+		return false;
+	}
+
+	if (schema instanceof ObjectNodeStoredSchema) {
+		if (node.fields.size !== schema.objectNodeFields.size) {
+			return false;
+		}
+		for (const [fieldKey, field] of node.fields) {
+			const fieldSchema = schema.objectNodeFields.get(fieldKey);
+			if (
+				fieldSchema === undefined ||
+				!isFieldInSchema(field, fieldSchema, nodeSchemaCollection, schemaPolicy)
+			) {
+				return false;
+			}
+		}
+	}
+
+	if (schema instanceof MapNodeStoredSchema) {
+		for (const field of node.fields.values()) {
```

Comment:
One thing that theoretically should be included here for map schema is:
if (complyWithMultiplicity(0, kind.multiplicity) {
// Map nodes implicitly have an infinite number of empty fields (for all the unused keys).
// Therefor the map must permit the empty case.
return false;
}